core/eolearn/tests/test_utils/test_parsing.py

```diff
@@ -0,0 +1,262 @@
+            (FeatureType.DATA, "CLM"),
+            (FeatureType.DATA, "BBOX"),
+            (FeatureType.MASK, "bands"),
+            (FeatureType.MASK, "CLP"),
+        ],
+    ),
+]
+
+TEST_CASES_ALLOWED = [
+    TestClass(
+        input=(
+            (
+                (FeatureType.DATA, "bands", "new_bands"),
+                (FeatureType.MASK, "bands", "new_bands"),
+                (FeatureType.MASK, "CLP", "new_CLP"),
+            ),
+            (FeatureType.MASK),
+        ),
+        features=[],
+        renaming=[],
+        specifications=[],
+    ),
+]
+
+
+@pytest.mark.parametrize("test_case", TEST_CASES_EMPTY + SPECIAL_CASES + TEST_CASES)
+def test_FeatureParser(test_case: TestClass):
+    parser = FeatureParser(test_case.input)
+    assert parser.get_features() == test_case.features
+    assert parser.get_renamed_features() == test_case.renaming
+    assert parser.get_feature_specifications() == test_case.specifications
+
+
+@pytest.mark.parametrize("test_case", TEST_CASES_ELLIPSIS)
+def test_FeatureParser_Error(test_case: TestClass):
+    """failing by design because ... does not have meaning without EOPatch"""
+    parser = FeatureParser(test_case.input)
+    with pytest.raises(ValueError):
+        parser.get_features()
+    with pytest.raises(ValueError):
+        parser.get_renamed_features()
+
+
+@pytest.fixture(name="eopatch")
+def eopatch_fixture():
+    eopatch = EOPatch()
+    eopatch.data["bands"] = np.arange(2 * 3 * 3 * 2).reshape(2, 3, 3, 2)
+    eopatch.data["IS_VALID"] = np.arange(2 * 3 * 3 * 2).reshape(2, 3, 3, 2)
+    eopatch.data["CLP"] = np.arange(2 * 3 * 3 * 2).reshape(2, 3, 3, 2)
+    eopatch.data["CLM"] = np.arange(2 * 3 * 3 * 2).reshape(2, 3, 3, 2)
+    eopatch.data["BBOX"] = np.ones((5, 4, 5, 2))
+    eopatch.mask["bands"] = np.arange(2 * 3 * 3 * 2).reshape(2, 3, 3, 2)
+    eopatch.mask["IS_VALID"] = np.arange(2 * 3 * 3 * 2).reshape(2, 3, 3, 2)
+    eopatch.mask["CLP"] = np.arange(2 * 3 * 3 * 2).reshape(2, 3, 3, 2)
+    eopatch.mask["CLM"] = np.arange(2 * 3 * 3 * 2).reshape(2, 3, 3, 2)
+    return eopatch
+
+
+@pytest.mark.parametrize("test_case", TEST_CASES_EMPTY + TEST_CASES + SPECIAL_CASES + TEST_CASES_ELLIPSIS)
+def test_FeatureParser_EOPatch(test_case: TestClass, eopatch: EOPatch):
+    parser = FeatureParser(test_case.input)
+    assert parser.get_features(eopatch) == test_case.features
+    assert parser.get_renamed_features(eopatch) == test_case.renaming
+
+
+@pytest.fixture(name="empty_eopatch")
+def empty_eopatch_fixture():
+    eopatch = EOPatch()
+    eopatch.data["CLP"] = np.arange(2 * 3 * 3 * 2).reshape(2, 3, 3, 2)
+    eopatch.mask["bands"] = np.arange(2 * 3 * 3 * 2).reshape(2, 3, 3, 2)
+    return eopatch
+
+
+@pytest.mark.parametrize("test_case", TEST_CASES + TEST_CASES_ELLIPSIS)
+def test_FeatureParser_EOPatch_Error(test_case: TestClass, empty_eopatch: EOPatch):
+    """failing by design because the test_case.input is not subset of EOPatch attributes"""
```

Comment:
docstrings should be grammatically correct. This includes upper case letters at the beginning and even possibly punctuation at the end

Reply:
I am not sure